### PR TITLE
3.9: Scripting

### PR DESCRIPTION
**Pull request by Matyrobbrt**

### CODE REVIEW

**Review by Matyrobbrt — COMMENTED** [review]
Everything is fine

File: src/main/java/com/mcmoddev/mmdbot/modules/commands/community/CmdEvaluate.java [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+        aliases = new String[] {"eval"};
+        help = "Evaluates the given script";
+        options = List.of(new OptionData(OptionType.STRING, "script", "The script to evaluate.").setRequired(true));
+    }
+
+    @Override
+    protected void execute(final SlashCommandEvent event) {
+        event.deferReply().queue(hook -> {
+            try {
+                ScriptingUtils.evaluate(Utils.getOrEmpty(event, "script"), createContext(new EvaluationContext() {
```
Comment (TheCurle) [inline src/main/java/com/mcmoddev/mmdbot/modules/commands/community/CmdEvaluate.java]: Would it not be a good idea to cache these contexts? Creating a new one and discarding it every time a slash command is invoked seems like a not-good idea.

File: src/main/java/com/mcmoddev/mmdbot/modules/commands/community/contextmenu/message/ContextMenuGist.java [hunk 2]
```diff
@@ -95,11 +96,16 @@ public static String generateName(int length) {
             .toString();
     }
 
+    public static final List<String> BLACKLISTED_ATTACHEMENTS = List.of("png", "jpg", "jpeg", "webm");
```
Comment (TheCurle) [inline src/main/java/com/mcmoddev/mmdbot/modules/commands/community/contextmenu/message/ContextMenuGist.java]: Probably worth cross-referencing the list of MAME multimedia types here - HEVC and mp4 will probably show up a lot in MMD.

File: src/main/java/com/mcmoddev/mmdbot/utilities/dictionary/DictionaryUtils.java [hunk 3]
```diff
@@ -42,6 +43,10 @@
 
     private static final HttpClient CLIENT = HttpClient.newHttpClient();
 
+    public static final DictionaryEntry CURLE_ENTRY = new DictionaryEntry("curle", "kəːle",
+        List.of(new DictionaryDefinition("noun", "A seemingly time travelling computer woman with a technological curse, a time-travelling penny, and no fucks to give whatsoever",
```
Comment (TheCurle) [inline src/main/java/com/mcmoddev/mmdbot/utilities/dictionary/DictionaryUtils.java]: Time travelling appears twice here. Probably worth changing "time-travelling penny" to "teleporting penny" just for the flow of the joke, even if it's not exactly what sci originally said :P

**Review by TheCurle — APPROVED** [review]
It's clear that a lot of effort has been put into this.
Without being able to test the scripting portions (due to my computer situation), i can't verify how well it works, so i trust your testing process.

That being said, i do like the contexts and items that you expose to the scripting engine.
You should think about how this can be abused, though:

- Would a user be able to use a script trick to spam a message in a channel?
- Could you potentially get a list of channels in the guild? That'd be a bad idea.
- Does the script expose the ability to embed images?

Other than those thoughts, this looks good to me.

File: src/main/java/com/mcmoddev/mmdbot/modules/commands/community/CmdEvaluate.java (hunk 1 above)
Comment (Matyrobbrt) [reply src/main/java/com/mcmoddev/mmdbot/modules/commands/community/CmdEvaluate.java]: That is not possible, due to the fact that each context has "bindings" (which contain all of the context members: functions, classes, variables), and those bindings cannot be discarded.

With that being said, contexts are designed to be created for every evaluation. And such, the first evaluation may take 2~3 seconds, but after that the instantiation is almost instant.

**Comment by Matyrobbrt** [comment]
> Could you potentially get a list of channels in the guild?

Yes, _but_ you can only get the channel's name and ID, and you cannot send anything in it. You can only send messages in the channel the evaluation was executed.

> Does the script expose the ability to embed images?

If you mean "embed image" by making the bot send a link yes. But making an embed is not possible unless you have the needed perms.

> Would a user be able to use a script trick to spam a message in a channel?

This is a good point, so I should actually timeout the evaluation after 4 seconds

**Pull request merged**